### ISSUE TITLE
[Enhancement] Sort Table Columns

### ISSUE THREAD

**Issue by robinsiebler** [beeware/toga#954]
## Expected Behavior
When you click on a column header, it should sort on the column by ascending/descending.

## Current Behavior
You can click on a column header, but nothing happens.

## Steps to reproduce
1. Create an app with a table

2. Populate the table

3. Run the app

4. Click on a column header

## Your Environment

* Python Version (list the specific version number): 3.7.6

* Operating System and Version (select from the following and list the specific version number; if your OS is not listed, list that as well)

  - [x] macOS - version: 10.15.5
  - [ ] Linux - distro:   - version:
  - [ ] Windows - version:
  - [ ] Other - name:  - version:

* Toga Version (list the specific version number or git hash)
0.2.15
* Toga Target (the type of app you are trying to generate)

  - [ ] android
  - [x] cocoa
  - [ ] django
  - [ ] gtk
  - [ ] iOS
  - [ ] tvOS
  - [ ] watchOS
  - [ ] winforms
  - [ ] win32
  - [ ] Other (please specify)

**Comment by freakboy3742** [comment]
Definitely a good feature to add; it will require integration with the data source to make this work.

**Comment by SimonVerhoek** [comment]
Just ran into this feature missing as well.

How hard would it be to simply add a `on_header_select` parameter to Table?

**Comment by freakboy3742** [comment]
Just adding a handler shouldn't be *that* difficult. Connecting that handler to something meaningful is the complicated part - which means, at the very least, implementing a sort mechanism on data sources.

**Comment by Oliver-Leigh** [comment]
I've implemented this for Windows at https://github.com/Oliver-Leigh/toga/tree/table-sorter

**Comment by freakboy3742** [comment]
@Oliver-Leigh Nice! I've got some thoughts about the API and approach, but it looks like you've got the main details worked out. 

It would be great if this feature was added to Toga as a feature for everyone to use; we'd need to work through those design issues. But if you're up for that discussion, we'd love to see the contribution.

**Comment by Oliver-Leigh** [comment]
@freakboy3742 That sounds good! I’ve only just started work on the macOS version, so it will be good to make the changes before I progress too much. 

What are your thoughts on what needs changing?

**Comment by freakboy3742** [comment]
My main concern is the high level separation of concerns: is sorting the responsibility of the source, or the widget? 

As an example of the problem: If I have 2 tables pointing at the same source, and I change the sort order in the first table - does the second table change sort order? Based on the implementation you've got here, the answer would be yes - but that feels like it might be unexpected behavior.

It seems to me that sorting should probably be a "mapping" behavior that is specific to a particular widget - something that can be put between a source and the widget as a proxy for row lookup, so if the widget asks for "row 3", that's a request for a *visual* row, which is mapped to an index in the *logical* data. The default mapping is `lambda x:x` (i.e., passthrough); any time you get a notification about a new data item, the sort mapping needs to be updated to reflect the new visual->logical mapping.

Does that make sense?

**Comment by Oliver-Leigh** [comment]
That makes sense! I hadn't thought about the use case of multiple widgets accessing the same data source. 

In regards to whether the sorting the responsibility of the source or the widget, I had been leaning towards the source. Mainly because this aligns with the native list sorting in python.

A solution could be to keep the sorting functionality as a method of the ListSource class, but as a "sort_mapping" method which returns the mapping. This way we wouldn't have to implement the same sorting functionality in in every widget. What do you think?

**Comment by freakboy3742** [comment]
That approach sounds like it could work - the ListSource then knows *how* to sort, and can create a representation that is optimal for whatever the implementation is. For example, if the ListSource is backed by a web API, it might make sense (or be necessary) to call the API to get sorted data, rather than call `sort(data)` in Python.

It would also mean that if you have 2 tables looking at the same source, both sorted by Name, you only need to maintain one sorted mapping. In the case where you *know* you're only going to have 1 view of the data, you *could* sort the original data source in place. (That wouldn't be a great idea architecturally, but it would be *possible*).

So - I guess we're looking at an API that looks something like:
* `can_sort(accessor: str) -> bool` - return if a source *can* be sorted by that column.
* `sorted(accessor: str) -> SortMapping` return a mapping object that converts sorted index to raw index

We could maybe reduce this to a single `sorted()` method that raises a `NotSortable` exception if the column isn't sortable. Not sure if there's an API need to independently evaluate sortability. Or I guess we could do both.

The other thought is whether `sorted()` should return a SortMapping which just provides a simple sorted-row-to-original-row translation mechanism, or something that fully adheres to *ListSource*. If we return a full *source*, then changing sorting is "change the source"; and in the trivial case, `sorted()` can return `self`. If you ask for a *different* sorting, the SortedListSource would need to retain a reference to the original, so it would proxy that call to sort the original.

The only other thing I want to keep in mind is the overlap with TreeSource. Whatever API we use for ListSource should have a reasonably close analog with TreeSource. I'm fairly certain both approaches would work; although how the sorting API is implemented on Node could be interesting. 

I'd be interested to know if @corranwebster has any thoughts on this, given his background with Qt's table/tree mechanisms (and his recent wrestling with Table/Tree primitives)

**Comment by Oliver-Leigh** [comment]
I'm leaning towards returning a simple list of indices. Then the widget uses this mapping. This seems like the most lightweight solution. However, a potentially messy part would be getting the widget to request a new mapping when there are changes to the table. 

I propose for the `sorted()` (or maybe `sort_mapping()`?) method to have the following arguments: 
 - `col_order: list[int] or None = None` - This gives the sort order and direction for the columns. E.g. something like [3,-2,-1,4]. (None defaults to [1,2,3,4].)
- `col_keys: Dict[int, Callable[...,Any]] or None = None` - Functions that tell the sorter how to sort given columns. This can also be used by the widget to make a given column sortable. 

The `can_sort` question is a little tricky because we also have the `missing_value` functionality of the table widget. So a column which contains `str` and `None`, could potentially just contain `str` in the Table. In my original implementation I had a `missing_value` argument in the sort function, but this is very Table specific and feels wrong to me now. This could be emulated with the `col_keys` argument that I proposed above. 

What I'm thinking is to add a `data_types` attribute to ListSource. This is a `list` (or maybe `dict`) of sets which contain the type(s) of each column. This gets set when `__init__` is run and updated when rows/columns are added/removed. The attribute easily answers the sortability question, but also gives richer information.

**Comment by freakboy3742** [comment]
> I'm leaning towards returning a simple list of indices. Then the widget uses this mapping. This seems like the most lightweight solution. However, a potentially messy part would be getting the widget to request a new mapping when there are changes to the table.

It shouldn't be *that* messy - there's already notification mechanism in place when data is added, removed, or changed; the sorting mechanism needs to re-evaluate whenever any notification is received. 

There is a notification *ordering* issue - we need to ensure that the mapping is re-sorted/evaluated before any GUI representation is updated. This is one of the reasons I think having a full proxied Source would make sense - that way, the sorted mapping listens to the original source, and the widget listens on the sorted mapping. Changing sort order means changing the data source to a new sorted proxy.

It's also worth pointing out the overlap with the work that @corranwebster has been doing recently on #3760; in this view of the world, a SortedListSourceT would be a readonly view of a ListSourceT.

> I propose for the `sorted()` (or maybe `sort_mapping()`?) method to have the following arguments:
> 
> * `col_order: list[int] or None = None` - This gives the sort order and direction for the columns. E.g. something like [3,-2,-1,4]. (None defaults to [1,2,3,4].)
> * `col_keys: Dict[int, Callable[...,Any]] or None = None` - Functions that tell the sorter how to sort given columns. This can also be used by the widget to make a given column sortable.

I'm not sure about this - or, at least, it depends on where these keys are specified. 

There are two possible views on sorting that I can see. Either:

1. Sorting is an inherent property of a specific ListSource. You ask the source to sort on a specific attribute, and you get data sorted on that attribute.
2. Sorting is a property of the Table. When a user clicks on a specific column, that defines a sort order that needs to be applied. 

It looks like you're broadly pointing towards option 2 (API specifics notwithstanding). Option (2) is definitely the most generic option; we can leverage all of Python's existing language and expectations around providing a `key` for sorting. A specific Column in a Table would have a sorting key; the mapping mechanism would be almost entirely generic "index for index, given a specific sorting key". It also means two different table instances could have a different mechanism for sorting on a given column.

However, that approach precludes any "smarts" in the implementation. In the "simple" case of using a Toga-provided ListSource, that doesn't really matter; but if you have an *independent* implementation of ListSourceT (that is, say, backed by a REST API), you *don't* have the ability to explicitly sort by an arbitrary key (unless you download all the data and do the sort locally).

So - I'm inclined to suggest that sorting is a property of the *source*, not the table. The table is a view *of* the source, and it might be possible to sort the data on a given column (or not); but you ask the source to provide a sorted view of the data on column X (or maybe a list of columns X, Y, Z); but how that is interpreted is up to the source. That means `col_order` is an argument passed to `ListSourceT.sorted(col_order)`; but ListSource itself needs to take the information you've defined as `col_keys`).

> The `can_sort` question is a little tricky because we also have the `missing_value` functionality of the table widget. So a column which contains `str` and `None`, could potentially just contain `str` in the Table. In my original implementation I had a `missing_value` argument in the sort function, but this is very Table specific and feels wrong to me now. This could be emulated with the `col_keys` argument that I proposed above.

I'm not sure I follow the problem here - I wouldn't have thought the missing value was involved in sorting at all. If the data are `[1,2,None,3]`, and missing value is 99, the data to be sorted isn't `[1,2,99,3]`. Missing value is entirely a presentation concern. 

`can_sort` is an entirely different issue. To me `can_sort` is controlling whether the UI should provide sorting as an option.
 
> What I'm thinking is to add a `data_types` attribute to ListSource. This is a `list` (or maybe `dict`) of sets which contain the type(s) of each column. This gets set when `__init__` is run and updated when rows/columns are added/removed. The attribute easily answers the sortability question, but also gives richer information.

I definitely don't follow why this is needed. Data type shouldn't affect sorting at all. Again, python has very well defined sorting semantics; we should be leaning on those as much as possible.

**Comment by Oliver-Leigh** [comment]
> So - I'm inclined to suggest that sorting is a property of the _source_, not the table. The table is a view _of_ the source, and it might be possible to sort the data on a given column (or not); but you ask the source to provide a sorted view of the data on column X (or maybe a list of columns X, Y, Z); but how that is interpreted is up to the source. That means `col_order` is an argument passed to `ListSourceT.sorted(col_order)`; but ListSource itself needs to take the information you've defined as `col_keys`).

Actually, I think that this is exactly the approach that I'm trying to convey!


> I'm not sure I follow the problem here - I wouldn't have thought the missing value was involved in sorting at all. If the data are `[1,2,None,3]`, and missing value is 99, the data to be sorted isn't `[1,2,99,3]`. Missing value is entirely a presentation concern.

I think we have few options here: 

1.  Block sorting for `[1,2,None,3]`, or
2.  Sort this as `[1,2,missing_value,3]` (provided the types align). 
3.  Sort this via the string representation. 

In my original approach I was trying to make things as flexible as possible, so I went with option 2 and then option 3 as a fallback. However, I'm completely fine with approach 1!


> I definitely don't follow why this is needed. Data type shouldn't affect sorting at all. Again, python has very well defined sorting semantics; we should be leaning on those as much as possible.

This was to help with option 2 (above). My idea is to get the sortability from the data types. That way we can see whether `missing_value` fixes the sortability issue. For example `{int,None}` becomes `{int}` which is now sortable. 

Lastly, my suggestion for a `col_keys` argument was also to help with option 2. If we're going with option 1, this is not needed. (I still think a full list of column orderings would nice because it would be useful if we wanted to implement multicolumn sorting down the line. Maybe a `int` or `list[int]` type would be nice?)

**Comment by corranwebster** [comment]
My gut feel is roughly this:
- the table widgets potentially have ways of getting the user to specify *how* to sort (eg. sort icons in column headers)
- you might want to give the user other ways to specify a sort than via the table widget (eg. buttons, menu items, etc.)
- sources have a way of saying that they can be sorted, and table widgets can hook into this
- if the source is sortable, and you share the source, then you share the sorting

Given that we're starting in a world where people have already written data sources without support for sorting, I'm tempted to say that sortability has to be detected by something like whether or not the data source has a `sort` method rather than asking all existing source authors to go and add a `can_sort` flag which is always `False` (although it might be OK to add this to the base `Source` class, it's a bit outside that class's scope which is all about notifications).

I'm in two minds about the API:
- the simple approach is that sortable data sources have a `sort` method which is just like the standard Python `sort`/`sorted` API: you have a `key` function that takes whatever the "rows" are and sorts them using that key function. This is powerful, simple, and familiar.
- the other approach is something higher-level that permits things like doing a sort via a database query, since that has the potential to be much more efficient either in time or memory usage.

However, the second approach makes it much harder to do some of the things being discussed in #4041. If you have a synthetic column which is a sum of other columns, for example, how do you sort by that? Or even without the concepts of #4041, what if you just had a proxy data source that did the addition and presented a `sum` accessor that the table can use, how does it pass through a request to sort by that value? Other issues come up around handling locale-aware sorts of text values using `locale.strxfrm`?

So much as I *like* the idea of being able to write super-smart Sources that perform database queries sorted by direct input from the UI, that sort of querying is "business/domain logic" in most cases, and not what a GUI layer should be worrying about directly. So we can presume that while the data sources that are being displayed in tables may be big, that they fit comfortably in memory and can be sorted in sub-second times by Python.

Which makes me lean towards the first option: sortable sources provide a `sort(key=..., reverse=...)` API that (ideally) does a stable sort on their data (so sorting by multiple columns can be done as a series of sorts, with highest priority last).

To handle all the issues around missing data, columns with incomparable data types, etc. we could have the `ColumnT` API from #4041 grow a `sort_key(row)` method, and then (given a stable sort in the data source) if the user clicks on a column to sort by it, that translates to a `data.sort(key=column.sort_key)` and then once the sort is done, the data source notifies that the data has changed, and the table updates appropriately.

So for example, if you had a column of numbers and missing values represented by None which you want sorted to the end, then you could have a custom column with a `sort_key` function that looks like:
``` python
class NumericColumn(AccessorColumn):
    def sort_key(self, row):
        value = self.value(row)
        return (1, 0) if value is None else return (0, value)
```

And given the existence of non-sortable data sources and the likely desire to have the ability to share underlying data, but sorted differently, we do want *proxy* data sources which do sorting (or make such sources really easy to write).  And the default `ListSource` class should grow a `sort` method which just sorts the internal list of rows in-place and fires an 'all updated' notification.

For Tree widgets, much the same idea, but the "nodes" also need to express the same sort API, and sorting becomes recursively sorting the children of each node.

The other wrinkle is that when it comes to the API presented for mutability, the data sources have largely been looking like the `MutableSequence` protocol with `__setitem__`, `__delitem__`, `insert`, and `append` by index; where for a sorted data source the API has more of `MutableSet` feel: you want to `add` and `discard` items by value without having to find which row index they are in and let the source sort them into the right place (although not every data source will be like this).

**Comment by Oliver-Leigh** [comment]
> * the other approach is something higher-level that permits things like doing a sort via a database query, since that has the potential to be much more efficient either in time or memory usage.

One possibility here is to rewrite the "backend" of ListSource to store the data via a polars dataframe. It natively stores most data that users would want and has an object wrapper for other python objects. There is some wrangling needed with the row object, and sorting the non-polars object types. However, I’ve worked out the logic, and I think it can be written as a drop-in replacement. 



> And given the existence of non-sortable data sources and the likely desire to have the ability to share underlying data, but sorted differently, we do want _proxy_ data sources which do sorting (or make such sources really easy to write).

Could you explain this a bit more? Would a proxy data source be much different to the mapping that @freakboy3742 was suggesting? It seems like there would be significant overlap here.

**Comment by corranwebster** [comment]
> One possibility here is to rewrite the "backend" of ListSource to store the data via a polars dataframe. It natively stores most data that users would want and has an object wrapper for other python objects. There is some wrangling needed with the row object, and sorting the non-polars object types. However, I’ve worked out the logic, and I think it can be written as a drop-in replacement.

I come from a scientific Python background, and I'd hesitate to make Toga depend on Polars - it's a Rust-based extension, and it's not entirely clear to me that it has pre-built wheels for iOS and Android (it *may*, it's just that the packaging structure is opaque on initial inspection). It's also "new and shiny" and has to prove its longevity. Finally, while data frame-like structures are great for Tables, sometimes they aren't a good fit for Trees.

I'd have most of the same hesitation about Numpy, Pandas, etc. even though there have been a few times since I started seriously working on Toga that I wished I'd had Numpy available.

OTOH a 3rd-party package that provided a Polars-based Source with all the bells and whistles would be cool.

> Could you explain this a bit more? Would a proxy data source be much different to the mapping that @freakboy3742 was suggesting? It seems like there would be significant overlap here.

Yes, it's more or less what @freakboy3742 was talking about: the "proxy" terminology comes from Qt (eg. [SortFilterProxyModel](https://doc.qt.io/qt-6/qsortfilterproxymodel.html). It's just a Source that wraps another Source and does transformations on the underlying data, but I didn't want to limit the concept to just proxies which do sorting.

**Comment by Oliver-Leigh** [comment]
> I come from a scientific Python background, and I'd hesitate to make Toga depend on Polars - it's a Rust-based extension, and it's not entirely clear to me that it has pre-built wheels for iOS and Android (it _may_, it's just that the packaging structure is opaque on initial inspection). It's also "new and shiny" and has to prove its longevity. Finally, while data frame-like structures are great for Tables, sometimes they aren't a good fit for Trees.
> 
> I'd have most of the same hesitation about Numpy, Pandas, etc. even though there have been a few times since I started seriously working on Toga that I wished I'd had Numpy available.
> 
> OTOH a 3rd-party package that provided a Polars-based Source with all the bells and whistles would be cool.

I see! All good points. Perhaps this is a project for other time... 



> Yes, it's more or less what [@freakboy3742](https://github.com/freakboy3742) was talking about: the "proxy" terminology comes from Qt (eg. [SortFilterProxyModel](https://doc.qt.io/qt-6/qsortfilterproxymodel.html). It's just a Source that wraps another Source and does transformations on the underlying data, but I didn't want to limit the concept to just proxies which do sorting.

Great! I think we're getting a more solid picture of what's needed: 

- A `ListSourceProxy` class which is a wrapper for `ListSource`. 
  - The `sort(...)` method will ask `ListSource` for a "sort mapping" which it then stores as an attribute. 
  - The existing public methods will pass through to `ListSource` (via the sort mapping where needed).

Does that sound good? 

I had a couple questions though: 

- I'm not sure how to handle the `insert(self, index, data)` method for `ListSourceProxy`. Specifically, where would we insert the data in the original ListSource? I'm leaning towards: 
  - If the sort mapping is the identity, then do the obvious, otherwise put it at the end. 
- Would a `ListSourceProxy` object also be consider as a `ListSourceT` object? I would think yes, but maybe I've missed something. 
- Should the method in `ListSource` which does the sorting be called `sort(...)`? Should it have an argument of the form `return_as_mapping=False`?

**Comment by freakboy3742** [comment]
> > I come from a scientific Python background, and I'd hesitate to make Toga depend on Polars

FWIW - I'd have the same reservations. Toga doesn't even have a hard dependency on *Pillow*, and it's a lot easier to make the argument that Pillow is important to app development than Numpy, Pandas or Polars. 

That said - I *am* very much in favour of adding any sort of interface or plugin system that would make a seamless integration with data science packages possible.

> > Yes, it's more or less what [@freakboy3742](https://github.com/freakboy3742) was talking about: the "proxy" terminology comes from Qt (eg. [SortFilterProxyModel](https://doc.qt.io/qt-6/qsortfilterproxymodel.html). It's just a Source that wraps another Source and does transformations on the underlying data, but I didn't want to limit the concept to just proxies which do sorting.
> 
> Great! I think we're getting a more solid picture of what's needed:
> 
> * A `ListSourceProxy` class which is a wrapper for `ListSource`.
>   
>   * The `sort(...)` method will ask `ListSource` for a "sort mapping" which it then stores as an attribute.
>   * The existing public methods will pass through to `ListSource` (via the sort mapping where needed).
> 
> Does that sound good?

That sounds about right.
 
> I had a couple questions though:
> 
> * I'm not sure how to handle the `insert(self, index, data)` method for `ListSourceProxy`. Specifically, where would we insert the data in the original ListSource? I'm leaning towards:

Is there any reason `index` wouldn't go through the same mapping? That is, if I want to insert at row 3 on the *sorted* proxy, that converts to row 5 in the base data source, and we insert there?

> * Would a `ListSourceProxy` object also be consider as a `ListSourceT` object? I would think yes, but maybe I've missed something.

Yes - that's sort of the point. 

> * Should the method in `ListSource` which does the sorting be called `sort(...)`? Should it have an argument of the form `return_as_mapping=False`?

The exact name is getting into [bike shed](https://bikeshed.org) territory, but my inclination is that `sorted()` fits better because `sort()` implies that the target object will *become* sorted, rather than returning a sorted proxy.

As for `return_as_mapping`... I'm not sure I understand what that argument would be for.

**Comment by freakboy3742** [comment]
> My gut feel is roughly this:
> 
> * the table widgets potentially have ways of getting the user to specify _how_ to sort (eg. sort icons in column headers)
> * you might want to give the user other ways to specify a sort than via the table widget (eg. buttons, menu items, etc.)
> * sources have a way of saying that they can be sorted, and table widgets can hook into this
> * if the source is sortable, and you share the source, then you share the sorting

Agreed on this as a set of principles.
 
> Given that we're starting in a world where people have already written data sources without support for sorting, I'm tempted to say that sortability has to be detected by something like whether or not the data source has a `sort` method rather than asking all existing source authors to go and add a `can_sort` flag which is always `False` (although it might be OK to add this to the base `Source` class, it's a bit outside that class's scope which is all about notifications).

+1 to this as well. A bare bones custom Source without a sorting implementation should still be usable on Table - but none of the columns would be sortable.
 
> I'm in two minds about the API:
> 
> * the simple approach is that sortable data sources have a `sort` method which is just like the standard Python `sort`/`sorted` API: you have a `key` function that takes whatever the "rows" are and sorts them using that key function. This is powerful, simple, and familiar.
> * the other approach is something higher-level that permits things like doing a sort via a database query, since that has the potential to be much more efficient either in time or memory usage.
> 
> However, the second approach makes it much harder to do some of the things being discussed in [#4041](https://github.com/beeware/toga/issues/4041). If you have a synthetic column which is a sum of other columns, for example, how do you sort by that? Or even without the concepts of [#4041](https://github.com/beeware/toga/issues/4041), what if you just had a proxy data source that did the addition and presented a `sum` accessor that the table can use, how does it pass through a request to sort by that value? Other issues come up around handling locale-aware sorts of text values using `locale.strxfrm`?

I'd argue that's a detail for a specific implementation. 

As a general pattern - A Column asks the Source for a sorted view of the data, using some identifier.

A simple AccessorColumn accessing a ListSource asks the ListSource for a sorted view based on the accessor. 

A custom CompositeColumn asks the Source for a sorted view based on "x+y"... which the Source can then reject because "WTF is x+y?" - or it can *accept* because it's a SQL-backed source and can turn it into a complex SQL request.
 
So - the end game is that to have a column sort, you need:
1. a Column implementation that can provide a sort key, 
2. a Source that supports sorting
3. a Source that supports the specific sort key.
 
If any of those conditions aren't met, the column isn't sortable. 1 and 2 are conditions that are soft "meeting a sortable interface" conditions; 3 is a hard data validation condition.

> So much as I _like_ the idea of being able to write super-smart Sources that perform database queries sorted by direct input from the UI, that sort of querying is "business/domain logic" in most cases, and not what a GUI layer should be worrying about directly. 
>
> So we can presume that while the data sources that are being displayed in tables may be big, that they fit comfortably in memory and can be sorted in sub-second times by Python.

I'm not sure that's entirely accurate. One of the key design ideas behind using Sources in the first place was decoupling the . The "back of my head" use cases have always been things like:
* a filesystem browser (a set of OS calls listing available files, sorted by name, date, ...)
* an email client (list of emails in your inbox is an IMAP query, possibly sorted by date, thread id, ...)
* a browser of your AWS resources (a set of (paginated) AWS API calls, possibly sorted by resource name, id, ...)
* A SQL database table inspector (a set of SQL calls, sorted by column)

In all those cases, the data source may not be able to hold a *full* copy of the data. A SQL database with millions of rows might get an indication of how many rows there are, and retrieve "pages" of data; but the source wouldn't ever have a *full* image of that  data.

Yes, that's all in business logic - but the purpose of the separation is to allow the Source to hold that business logic, and let the GUI say "show me rows 500-600 of the available 2M rows of data", independent of how that request will be satisfied.

> The other wrinkle is that when it comes to the API presented for mutability, the data sources have largely been looking like the `MutableSequence` protocol with `__setitem__`, `__delitem__`, `insert`, and `append` by index; where for a sorted data source the API has more of `MutableSet` feel: you want to `add` and `discard` items by value without having to find which row index they are in and let the source sort them into the right place (although not every data source will be like this).

Agreed that mapping insert/delete operations into a sorted source is (at the very least) problematic. I don't think it's unreasonable to say that data *modification* APIs *can* need to be restricted to the upstream "original' source, as long as there's a documented mechanism to get from a sorted source back to it's original.

**Comment by mhsmith** [comment]
My first impression is that this all sounds unnecessarily complicated. Understanding the concept of sources is difficult enough without introducing extra layers of proxies and composites, and worrying about how they interact with listeners and modifications.

What about this:
* The source order is always the same as the display order.

* ListSource has a `sort` method identical to [that of `list`](https://docs.python.org/3/library/stdtypes.html#list.sort).

* Table has an event handler `on_sort(accessor, reverse)`, which does nothing by default. If it is implemented, then Toga will track the sorted column and direction, and indicate it in the UI if the native toolkit supports that.

* If the app is using ListSource, then it can enable sorting by implementing `on_sort` as follows:
  ```py
  from operator import attrgetter
  table.data.sort(key=attrgetter(accessor), reverse=reverse)
  ```

* If the app is using some other implementation of Source, then implementing `sort` is optional, just as implementing all the other modification methods is. For example, if the data comes from a web API request as mentioned [above](https://github.com/beeware/toga/issues/954#issuecomment-3751999086), then the developer should probably leave `sort` unimplemented, and initiate the request in the `on_sort` handler.

One scenario which I have in mind is using Table to display the results of a search which returns too many items to display at once. In this case, the Source doesn't represent the entire dataset, but only the subset of it which is currently displayed. The app would provide "next/previous page" buttons to load more data, and changing the sort order would not merely reorder the existing subset of items, but display a different subset.

**Comment by freakboy3742** [comment]
I agree that's a much simpler interface. The one major consequence is that it's impossible to have 2 views of the same source that have a different sort order. Taking the table_source example - you're either sorting *both* by year, or you're sorting *both* by movie name. You can't sort one by year, and one by movie name.

I guess it's reasonable to consider "multi-view" list sources to be the edge case, though. It should still be *possible* to build a proxied ListView - table_source is already an example of a proxy; building one that maintains a different sort order should be entirely possible. We might even want to provide a built-in proxy class for the build-in ListSource; but that wouldn't be a core part of the sorting API (and definitely isn't needed for the rest of this proposal to work).

> One scenario which I have in mind is using Table to display the results of a search which returns too many items to display at once.

That's an interesting use case - a search space with 10k results, but you retrieve 100 results per web query, and the UI only attempts to display 100 at a time, and then sorts on the *Python* side. 

I can see how that would be implemented with this design. The problem I see is that it precludes *server-side* sorting as an alternative. `ListSource.sort(key=...)` is definitely a Pythonic sorting interface, but server-side sorting won't usually be "python compatible". 

We could possibly address this by allowing `key` to be open ended, and up to the Source as to how the value is interpreted. The built-in ListSource would require a Python-style key callable; but a custom API-backed source could accept an identifier, a SQL snippet, or whatever else make sense under the circumstances.

If we do that - a #4091-style `Column` could include the "sort key" (whatever value it is) as part of the definition, and then apply that sort on the data source whenever the sorting UI is activated on that column.

In the case of @corranwebster's "CompositeColumn" merging two source attributes, there's coordination between the Source and Column required, but it should still work. If the underlying data is a ListSource, then it could be as simple as a callable key that returns a tuple of the two attributes; or it could be some sort of identifier for a SQL query that lists the two columns in sort order.

`on_sort` would return that sort key as the payload; if a table receives an `on_sort` notification and it has a column that uses the same sort key (evaluated by instance, so that callables are comparable) then the UI can respond making that column appear to be sorted. If there's no match, the table would appear as "unsorted".

Does that sound workable?

**Comment by Oliver-Leigh** [comment]
I don't think this needs to be complicated. What I believe @corranwebster was suggesting (correct me if I'm wrong) is that the `Table` widget sends a command `sort` that actually sorts the `ListSourceT` associated to the table (at least when the `ListSourceT` has a `sort` method).

The (potentially edge) case of  having a source with multiple views is covered by a new `ListSourceT` class called something like `ListSourceProxy`. In this case the user would have a base `ListSourceT` and `ListSourceProxy` wrapper around this, then the `Table` is constructed using the `ListSourceProxy` object. There are details to work out here e.g. Should the proxy do the sorting or should it request a sorting from the source? Should the proxy listen to source using the standard source listener? etc...

If we accept the above approach, I think there are actually 3 separate features being discuss here: 

 1. `Table` header sorting functionality
 2. A `ListSourceProxy` object.
 3. A pagination/sublist-display feature

On top of this, I think we need a few more features to make things work smoothly: 

 4. A selection setter for the `Table` widget. (Discussed in #2702)
 5. An `__iter__` method for ListSource. (This would be nice to have anyway.)

**Comment by corranwebster** [comment]
The last few comments made me realise I was missing something - I was focused on the mechanics of sorting and how a table tells a source to sort. But there is also the reverse problem: a source has been sorted (somehow, maybe as part of a database query), how can it signal to the UI that it is already sorted in a particular way so that the table can display a sort indicator in a header, or similar?

An `on_sort` event is only part of the story, there needs to be something that a table widget can look at during initialisation that holds the sort key information.

So how does the following sound, building on what @freakboy3742 suggested:
- sortable sources have a "sort_key" attribute that contains the information about how the source is sorted
- exactly what a `sort_key` is is left undefined, other than any two sort keys which define the same sort order should ideally be *equal* (probably not doable in general for synthetic columns - working out that sorting by `a + b` is the same as sorting by `b + a` would be way out of scope).
- following @freakboy3742's suggestion the Column interface grows a `sort_key` attribute which returns a sort key usable by the data source being viewed (this is different from what I was originally proposing, which was that `sort_key` was a method)
- if a Column's sort key matches the source's sort key, then the Column can tell the table to display a sort indicator, if relevant
- if a table has a way of specifying a way to sort by a column (like clicking on a header), the table sets the source's sort_key to the column's sort_key (or maybe call's the column's `sort()` method with the sort key, but I think it needs to be async).
- if the sort key is changed on the source, the source sorts itself, and an event is fired when the sort is done
- setting the sort key is an async operation, to allow querying to a remote source like a database without blocking

There doesn't need to be a public `sort` method on sortable Sources, and if there is it could just be a wrapper around setting the sort_key and waiting for it to be done.

Having the sort key on the source means that a sorted source can do things like inserting new items into the correct position without having to do a full re-sort.

This requires a bit of collaboration between Columns and Sources: if a sortable Source is using non-standard sort keys, then it needs non-standard Columns that understand its mechanics.

**Comment by Oliver-Leigh** [comment]
Is this just added complexity for not much benefit? 

Surely a simpler approach would be to have a method in the Table widget that implements a header click. On a header click, the table simply says to the `ListSourceT` object "_sort yourself with a given column priority and according to each column's data type_". It then updates itself accordingly.  This is the natural behavior I would expect from a header sort. 

Yes, a sort _can_ also be performed on the `ListSourceT` object directly using an arbitrarily complicated key, but _should_ this be reflected in the header of the table? As you point out, it is somewhat complicated to decided whether a sort can or should be displayed in the header. 

The only benefit I can see is that one could perform a different type of sort on the column. However, I think this is an edge case, and it can probably also be achieved by adding a wrapper to the column with different `__gt__`, `__lt__`, `__eq__` methods. 

Since the header indicates whether it's been clicked, I think the simplest approach is for the table widget to be responsible for choosing and displaying a header sort.

**Comment by Oliver-Leigh** [comment]
> I can see how that would be implemented with this design. The problem I see is that it precludes _server-side_ sorting as an alternative. `ListSource.sort(key=...)` is definitely a Pythonic sorting interface, but server-side sorting won't usually be "python compatible".

This is not necessarily true. For general sorting it shouldn't be a problem since your calling the source (server) directly, and for the sorting called by the `Table` widget you could make a translation using a "two_hot_encoding" to recover the sort order.  Here's a simple example:

```python
def two_hot_encoding(n: int):
    def f(i: int, j: int, value: int):
        if i==j: 
            return value
        return 0
    positive = [ [ f(i, j, 1) for i in range(n) ] for j in range(n) ]
    negative = [ [ f(i, j,-1) for i in range(n) ] for j in range(n) ]
    return positive + negative

def get_sort_order(sort_encoded_list: list[list[int]]):
    n = len(sort_encoded_list[0])
    
    sort_order: list[int] = []
    desc_list: list[bool] = []

    for i, row in enumerate(sort_encoded_list):
        try:
            index = row.index(1)
        except:
            pass
        try:
            index = row.index(-1)
        except:
            pass

        sort_order.append(index)
        desc_list.append(row[index]>0)

        if i == n-1:
            break
    return (sort_order, desc_list)

test_list = two_hot_encoding(3)
test_list.sort(key=(lambda x: [-x[2],x[0],-x[1]]))

print(get_sort_order(test_list))
```

**Comment by freakboy3742** [comment]
> The last few comments made me realise I was missing something - I was focused on the mechanics of sorting and how a table tells a source to sort. But there is also the reverse problem: a source has been sorted (somehow, maybe as part of a database query), how can it signal to the UI that it is already sorted in a particular way so that the table can display a sort indicator in a header, or similar?

That's a good point.
 
> * if a Column's sort key matches the source's sort key, then the Column can tell the table to display a sort indicator, if relevant
> * if a table has a way of specifying a way to sort by a column (like clicking on a header), the table sets the source's sort_key to the column's sort_key (or maybe call's the column's `sort()` method with the sort key, but I think it needs to be async).
> * if the sort key is changed on the source, the source sorts itself, and an event is fired when the sort is done

Is there a need here to be able to determine if a column *could* be sorted? If I'm understanding this proposal, `sort_key` would represent "the source is currently sorted by this key". If a column is going to display any sort of hint that it *can* be sorted, presumably we need to be able to interrogate that availability?

> * setting the sort key is an async operation, to allow querying to a remote source like a database without blocking

Agreed this is a concern, or at least needs to be an option. Strictly, I guess that is true of *all* methods (even things like insert or delete). 

However, it's also a situation where the async-ness could be in the background - since the "response" to a change on the source (whether that's sorting, insertion, or whatever) is a signal, a synchronous "sort()" method could *initiate* the sort, and when it's done, fire the `on_sort()`.
 
> There doesn't need to be a public `sort` method on sortable Sources, and if there is it could just be a wrapper around setting the sort_key and waiting for it to be done.

I'm not sure about this, for two reasons.

Firstly, from a practical perspective: how would reverse sorting work? Is that a different sort key? 

From a philosophical perspective: even if there's nothing else needed from an API perspective, sorting feels like something that is "imperative" - "this will have an effect", rather than changing a property of the source.

**Comment by freakboy3742** [comment]
> Is this just added complexity for not much benefit?
> 
> Surely a simpler approach would be to have a method in the Table widget that implements a header click. On a header click, the table simply says to the `ListSourceT` object "_sort yourself with a given column priority and according to each column's data type_". It then updates itself accordingly. This is the natural behavior I would expect from a header sort.

I'm not sure I see how it's more complex - isn't it the same thing? A Column *is* the definition of the header - saying a "header sorts with a given column priority" is the same as saying "a Column sorts with a given sort key".

> The only benefit I can see is that one could perform a different type of sort on the column. However, I think this is an edge case, and it can probably also be achieved by adding a wrapper to the column with different `__gt__`, `__lt__`, `__eq__` methods.

Implementations of `__gt__` et al are fine if the data is all *local*, but as noted in other comments - there are use cases where this *won't* be possible.
 

> > I can see how that would be implemented with this design. The problem I see is that it precludes _server-side_ sorting as an alternative. `ListSource.sort(key=...)` is definitely a Pythonic sorting interface, but server-side sorting won't usually be "python compatible".
> 
> This is not necessarily true. For general sorting it shouldn't be a problem since your calling the source (server) directly, and for the sorting called by the `Table` widget you could make a translation using a "two_hot_encoding" to recover the sort order. Here's a simple example:

I'm not sure what you're trying to describe here. The key insight about server-side sorting is that the Python/Toga side *does not* have access to *all* the data, and as a result, isn't in a position to do sorting. Any implementation based on invoking a sort key function on the data in the source will only work for the simple `ListSource` case, not the general `ListSourceT` case.

**Comment by Oliver-Leigh** [comment]
> I'm not sure I see how it's more complex - isn't it the same thing? A Column _is_ the definition of the header - saying a "header sorts with a given column priority" is the same as saying "a Column sorts with a given sort key".

It's not quite the same. For one, you can sort it different ways e.g. if column is type `int` you could also sort with `key = lambda x: str(x)`. 

My broader point is that you can also have sort keys which don't fit into the natural column structure. For example, if we have a `ListSourceT` with 2 columns we could have something unusual like: `key = lambda x: str(x[1])[:-5] + str(x[2]) + str(x[1])[-5:]`.

In either of these cases, we need `ListSource` to be able to determine whether it is appropriate for `sort_key`  to be reflected as a header click on the UI. But why bother? We already know that a sort request coming from the `Table` widget is exactly in the format we want to be reflected in the UI. I feel like we'd needlessly be adding complexity to the `ListSource` class, which (as @mhsmith points out above) is already very complex. 


> I'm not sure what you're trying to describe here. The key insight about server-side sorting is that the Python/Toga side _does not_ have access to _all_ the data, and as a result, isn't in a position to do sorting. Any implementation based on invoking a sort key function on the data in the source will only work for the simple `ListSource` case, not the general `ListSourceT` case.

I'm trying to describe how it could work for an interface/proxy to translate a `Table` sort key into the appropriate serve code. For example (continuing my example above) for a Polars server one could have: 

```python
sort_order, desc_list = get_sort_order(test_list)

df.sort(sort_order, descending=sort_desc)
```

For SQL server one could make a query as follows:
```python
def direction(desc:bool) -> str:
    if desc:
        return 'DESC'
    
    return 'ASC'

column_names = ["column1","column2","column3"]

sql_query = f"SELECT {", ".join(column_names)} FROM database_name ORDER BY "

for i, sort_index in enumerate(sort_order):
    sql_query += column_names[sort_index] + " " + direction(desc_list[i]) + ", "

sql_query = sql_query[:-2]
print(sql_query)
```

The interface/proxy would then pass the appropriate sort request to the server. Of course, this is implementation specific. I'm just trying to highlight a proof of concept.

**Comment by freakboy3742** [comment]
> My broader point is that you can also have sort keys which don't fit into the natural column structure. For example, if we have a `ListSourceT` with 2 columns we could have something unusual like: `key = lambda x: str(x[1])[:-5] + str(x[2]) + str(x[1])[-5:]`.

Sure - but I don't see why it's any more complex to define a Source that provides that sort key and a Column that references that sort key, vs a Column that defines the source key. The number of lines of code is almost identical; the only difference is where the sort key is being defined.
 
> The interface/proxy would then pass the appropriate sort request to the server. Of course, this is implementation specific. I'm just trying to highlight a proof of concept.

Ok - you've lost me here. In your proposal, what is the API contract for `sort()`? In every example you've given to date, it's a "classic python sort", where `key` is a lambda providing a sortable value; but I can't see how this SQL example would be implemented as that sort of key function.

**Comment by Oliver-Leigh** [comment]
> Sure - but I don't see why it's any more complex to define a Source that provides that sort key and a Column that references that sort key, vs a Column that defines the source key. The number of lines of code is almost identical; the only difference is where the sort key is being defined.

I agree, provided you've restricted yourself to a column specific key. My concern is determining whether a _general_ `sort_key` is appropriate to display in the column header. 



> Ok - you've lost me here. In your proposal, what is the API contract for `sort()`? In every example you've given to date, it's a "classic python sort", where `key` is a lambda providing a sortable value; but I can't see how this SQL example would be implemented as that sort of key function.

Can you elaborate a bit on your concern here? Once you have the SQL query, you can call it from python (using for example pyodc). I should stress that this is _only_ for a sort call coming from the table header. 

In my proposal, the `Table` would send a sort request via a lambda `key`, but as you point out this is python sort, so we need an interface. A priori, we know that this request is coming from a column order and asc/desc list, so one can utilize the encoding I provided to recover this column order and asc/desc list. Once you have this information you can construct a sort request in the native database language, as I demonstrate for SQL.

**Comment by freakboy3742** [comment]
> > Sure - but I don't see why it's any more complex to define a Source that provides that sort key and a Column that references that sort key, vs a Column that defines the source key. The number of lines of code is almost identical; the only difference is where the sort key is being defined.
> 
> I agree, provided you've restricted yourself to a column specific key. My concern is determining whether a _general_ `sort_key` is appropriate to display in the column header.

I'm confused by what you're describing here. `sort_key` *can't* be a display label if it's a lambda; why would it be if it's a sort label? It's a purely functional identifier: `Column(heading="The User's Name", accessor="username", sort_key="lexicographic")` would describe a column with a label of "The User's Name", drawing data from the `username` attribute of each row, performing sorting via the "lexicographic" sorting, with the source defining how "lexicographic" sorting was implemented. That could be a `lambda v: v.upper()` (to give a case insensitive sorting), or it could be a SQL query implementing server-side sort. It's up to the source to define what "lexicographic" sorting means - the text is nothing more than an identifier for a type of sorting that could be performed.

> > Ok - you've lost me here. In your proposal, what is the API contract for `sort()`? In every example you've given to date, it's a "classic python sort", where `key` is a lambda providing a sortable value; but I can't see how this SQL example would be implemented as that sort of key function.
> 
> Can you elaborate a bit on your concern here? Once you have the SQL query, you can call it from python (using for example pyodc). I should stress that this is _only_ for a sort call coming from the table header.
> 
> In my proposal, the `Table` would send a sort request via a lambda `key`, but as you point out this is python sort, so we need an interface. A priori, we know that this request is coming from a column order and asc/desc list, so one can utilize the encoding I provided to recover this column order and asc/desc list. Once you have this information you can construct a sort request in the native database language, as I demonstrate for SQL.

My concern is that I don't understand how you can define a server-side SQL call based on detail provided by a function that adheres to Python "sort_key" semantics. 

Python requires a `key` function to be lambda that takes a single row, and returns a sortable version of that row. Server-side sort means you *do not* have access to row-level detail at the time you're invoking a sort. A key function doesn't tell you the "name" of the column being sorted - it tells you how to extract the appropriate columns from *one* specific row.

**Comment by mhsmith** [comment]
> `ListSource.sort(key=...)` is definitely a Pythonic sorting interface, but server-side sorting won't usually be "python compatible".

My only reason for including a `ListSource.sort` method was to give an easy way to implement the simple case by treating the `ListSource` as a `list`. I agree that the `key` argument can't be implemented using a SQL query in general.

> That's an interesting use case - a search space with 10k results, but you retrieve 100 results per web query, and the UI only attempts to display 100 at a time, and then sorts on the _Python_ side.

That wouldn't be very useful, because the user probably wants to change the sort order of the entire query, which would usually require you to display a different set of 100 items.

In fact server-side sorting is exactly what I was thinking of with this example:
* The user clicks on a table header, which generates an `on_sort` event.
* The app responds to the event by generating a new query to the server. 
* The server does the sort and returns the top 100 items according to the new sort key.
* The app displays the data by creating a new source object or modifying the existing one. The source object does not do any sorting by itself, and does not need to implement any sort-related APIs. 

> But there is also the reverse problem: a source has been sorted (somehow, maybe as part of a database query), how can it signal to the UI that it is already sorted in a particular way so that the table can display a sort indicator in a header, or similar?

That's a good point, but again I think it's simpler to handle this at the level of the UI rather than of the source, and leave it up to the app to decide how to link the two. This could be done by giving `Table` the following properties:

* `sortable: bool = False` – whether the sort indicator can be interactively changed. Programmatic assignment of the other properties is still possible even when this is false.
  * I don't think per-column sortability is a common enough requirement to justify complicating the API. An app could still implement it by having its `on_sort` method ignore requests to sort by an invalid column, and reset the sort indicator to whatever it was before.
* `sort_column: str | None = None` – the accessor of the column with the sort indicator, if any.
  * I definitely don't think multi-key sorting is necessary, since there's no standard UI pattern to control it.
* `sort_reverse: bool = False` – the direction of the sort indicator.

`sort_column` and `sort_reverse` can be changed by the user clicking on the header, or programmatically. Either way, `on_sort` will then be called with no arguments, since it can access the new settings from the properties. This is consistent with the way we handle other interactively-modifiable properties.

**Comment by Oliver-Leigh** [comment]
@mhsmith I think this is might be the way to go! It's pretty close to what I've already written, so it shouldn't take too much reworking. The other features such as proxies and pagination could be worked out at a later date. 


> I'm confused by what you're describing here. `sort_key` _can't_ be a display label if it's a lambda; why would it be if it's a sort label? It's a purely functional identifier: `Column(heading="The User's Name", accessor="username", sort_key="lexicographic")` would describe a column with a label of "The User's Name", drawing data from the `username` attribute of each row, performing sorting via the "lexicographic" sorting, with the source defining how "lexicographic" sorting was implemented. That could be a `lambda v: v.upper()` (to give a case insensitive sorting), or it could be a SQL query implementing server-side sort. It's up to the source to define what "lexicographic" sorting means - the text is nothing more than an identifier for a type of sorting that could be performed.

Great! I think we’ve got to the bottom of this: I was thinking that sort_key was the general row _”sort key”_ and I was thinking this would cause a problem. You were already a few steps ahead.

> My concern is that I don't understand how you can define a server-side SQL call based on detail provided by a function that adheres to Python "sort_key" semantics.
> 
> Python requires a `key` function to be lambda that takes a single row, and returns a sortable version of that row. Server-side sort means you _do not_ have access to row-level detail at the time you're invoking a sort. A key function doesn't tell you the "name" of the column being sorted - it tells you how to extract the appropriate columns from _one_ specific row.

I think we might be talking a bit at cross purposes here, and we should probably revisit this at a later date since it's a side issue. I will be writing a polars-based ListSourceT for my own purposes. I’ve worked out the details, but haven’t written it yet. I think the issues you have will become clear once it is done.

**Comment by freakboy3742** [comment]
> > That's an interesting use case - a search space with 10k results, but you retrieve 100 results per web query, and the UI only attempts to display 100 at a time, and then sorts on the _Python_ side.
> 
> That wouldn't be very useful, because the user probably wants to change the sort order of the entire query, which would usually require you to display a different set of 100 items.

I mostly meant "interesting" in the "may you live in interesting times" sense :-) I agree it would be very odd design choice, and definitely not something we should optimize for.

> In fact server-side sorting is exactly what I was thinking of with this example:

In terms of this general workflow, I think we're in furious agreement.
 
>   * I don't think per-column sortability is a common enough requirement to justify complicating the API. An app could still implement it by having its `on_sort` method ignore requests to sort by an invalid column, and reset the sort indicator to whatever it was before.

Really? I would have thought it was an obvious and extremely likely requirement. It seems entirely reasonable (and likely) that if a table that has "name", "age", and "height" columns, a user might want to sort on any one of those three columns, in either direction.

> * `sort_column: str | None = None` – the accessor of the column with the sort indicator, if any.
>   
>   * I definitely don't think multi-key sorting is necessary, since there's no standard UI pattern to control it.

Agreed on this point; a user-facing API where you can specify "sort by A, then C, then B" *would* be an edge case, IMHO. 

(That doesn't prevent someone writing a sort mechanism for column A that *does* resolve ties by looking at C and then B, but that's down to the "how do I sort on A" implementation).

**Comment by freakboy3742** [comment]
> I think we might be talking a bit at cross purposes here, and we should probably revisit this at a later date since it's a side issue. I will be writing a polars-based ListSourceT for my own purposes. I’ve worked out the details, but haven’t written it yet. I think the issues you have will become clear once it is done.

I'm not sure I see how it's a "side issue" - it's a fundamental question of the API for sorting. The proposal @mhsmith has described seems to be predicated on "there is only one possible sort order for a ListSource", which is certainly a possibility - but it seems to miss the reason why users would want their Table data to be sortable.

**Comment by mhsmith** [comment]
> > * I don't think per-column sortability is a common enough requirement to justify complicating the API. 
> 
> Really? I would have thought it was an obvious and extremely likely requirement. It seems entirely reasonable (and likely) that if a table that has "name", "age", and "height" columns, a user might want to sort on any one of those three columns, in either direction.

I agree; I meant that the API should allow interactive sorting either by all columns, or none of them, hence the single boolean `sortable`.

> The proposal [@mhsmith](https://github.com/mhsmith) has described seems to be predicated on "there is only one possible sort order for a ListSource", which is certainly a possibility, but it seems to miss the reason why users would want their Table data to be sortable.

I would put it more as: the Table acccesses the source as a simple sequence; it doesn't have any API to ask it how it's sorted, any more than a standard `list` does. 
 
The source may have APIs for *doing* a sort, but maybe it's best to keep those implementation-specific. For example, the standard ListSource would have a `list`-like `sort` method that takes a `key` argument, but this would *not* be part of the ListSourceT API. Other source classes could have their own sort methods if appropriate, but since Toga itself would never call them, they wouldn't need to have the same argument format.

**Comment by Oliver-Leigh** [comment]
> The source may have APIs for _doing_ a sort, but maybe it's best to keep those implementation-specific. For example, the standard ListSource would have a `list`-like `sort` method that takes a `key` argument, but this would _not_ be part of the ListSourceT API. Other source classes could have their own sort methods if appropriate, but since Toga itself would never call them, they wouldn't need to have the same argument format.

I think this sound good. If someone wants a general non-python sort, then that can be implemented directly, maybe as a `native.sort()` method. However, the UI header sort uses the pythonic sort `key` approach of the built-in `ListSource`. Any other implementation would need an interface for the UI header sort. (How to construct this interface is what I've tried to describe above.)

**Comment by freakboy3742** [comment]
> I agree; I meant that the API should allow interactive sorting either by all columns, or none of them, hence the single boolean `sortable`.

*All* may not be either possible or desirable. For example, if a column contains a color definition... is sorting by RGB value really meaningful? Or consider the case of a "todo list" where items have a title, due date and a priority; you may want to be able to sort by priority and due date, but *not* by title. 

I guess I'm not sure I see the distinction between a boolean "This table can be sorted" returning False and "This table can be sorted by these columns, which is an empty list".

> I would put it more as: the Table acccesses the source as a simple sequence; it doesn't have any API to ask it how it's sorted, any more than a standard `list` does.

I'm not sure I see how that can work (or, at least, I'm missing a connection somewhere). 

In order to visualise sorting in the GUI, the Table viewing a source needs to:

* Know is possible to sort by column N, so it can respond, or not, to a sort request.
* Know if column N is (or is not) currently selected as the ascending (or descending) order, so it can display if a column is the current sort (if required by the GUI - eg, putting an up/down arrow in a column header)
* Allow the user to request that the table - and thus the source - be sorted by column N (in either order), in response to a click in a column header (or other UI/API request)
 
I agree that the actual sorting process should be implementation-specific, and should be implemented on the Source. But unless there's a API to do that sort on the source... what's the mechanism for requesting a change in sort order?

**Comment by corranwebster** [comment]
A quick note from my current dive into the Qt Table and Tree objects: if you want to preserve the selection (and expanded/collapsed nodes of a Tree) under sorting, Qt expects you to run through all the "persistent" nodes and map the old positions to the new positions. Haven't really thought through what this might imply in terms of the API, but thought I should note here while it's in my mind.

**Comment by freakboy3742** [comment]
I had a chance to catch up with @mhsmith in person last week, and we had a productive discussion that I think resolved some of the confusion around possibly approaches to column sorting.

The built in ListSource object is in a position to adopt a Pythonic `.sort(key, reverse)` API. However, that isn't a requirement that can be imposed on other data sources. This means that we either need a high level API that is agnostic of how sorting is implemented, or we need to move sorting out of the Source API (or at least make it optional).

An additional constraint that I hadn't fully considered was that when a data source requires a complex or server-side sort, it's entirely possible the sorting operation will need to be asynchronous, with the UI interaction *triggering* the sort, but not waiting until completion.

With these two things in mind, @mhsmith's suggestion is to completely separate the two concerns. Instead of trying to work out how to define a column so that it can tell a Source how to sort itself, we add an `on_sort()` handler to Table that is triggered whenever the user interacts with a column to select sorting. It's then up to the developer to define how to implement `on_sort()` - the simple version will be to call `self.data.sort(...)`, but any other implementation becomes possible.

This does give the flexibility that:
1. A column sort handler can call *any* API that may be needed to start the sort process (whether it's on the Source or not); and
2. The handler can be an async handler if required.

I agree these are both desirable properties. However, there are two notable downsides.

The first is that the trivial case of a ListSource with all (or even partially) sortable column types becomes moderately complex, as the developer needs to manually plumb all the sorting connections that could be otherwise implied. That's not a *complete* dealbreaker - but it does mean a more complex definition is required for the simple case. It might be possible to write a helper to simplify the trivial case (e.g., a ListSource.sort_handler())` method that generates the sort handler for the simple case.

The second, and more serious downside is that it's difficult to separate an implementation of `on_sort` from the source type. An implementation of `on_sort` that invokes `.sort(key=...)` on a ListSource will no longer work if a user changes to a different data source that satisfies the same accessor pattern, but doesn't provide a `.sort()` method. Any change in Source type will also potentially also require a change in the `on_sort` handler.

To me, this second point highlights that sorting *is* an inherently source-specific behaviour, so we should be encouraging the API to be *on* Source, via a "descriptive" approach to sorting - that is, a sort interface that can be implemented in different ways based on a high level description of the type of sort that has been requested. 

This approach doesn't preclude the handler being asynchronous; but it *does* allow the simple case to be one additional attribute on a Column definition (the descriptive sort property) - or even implied for the AttributeColumn case (although it might be better to be explicit).
